### ISSUE TITLE
OLD functions concept exercise

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#2353]
This issue describes how to implement the `functions` concept and concept exercise for the Python track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [What are Concept Exercises?](https://exercism.org/docs/building/product/concept-exercises)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts)
- [Concept Exercise Specifications](https://exercism.org/docs/building/tracks/concept-exercises) 
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

## Goal

This `functions` concept + concept exercise is meant to teach a deeper understanding and use of `functions`  in Python. It should also explain how Python treats/views functions (_as callable objects_).

## Learning Objectives

<details>
<summary><b>In Scope</b></summary>

- Understand more about  Python `scopes` and `namespaces`
  - understand the difference between the `global` and `nonlocal` keywords and when to use them
- Get familiar with the [special attributes](https://docs.python.org/3/reference/datamodel.html#types) of Python functions
- Get familiar with best practices when using `return`, and the difference between _explicit_ and _implicit_ `return`
   - functions without an explicit return will return the singleton object `None`
- Understand what is meant by _"functions are first class **objects** in Python"_.
  - understand that `functions` **are** `objects` in Python, and that they have `types`
  - understand that `functions` can be assigned to variables, used in expressions,  and stored in various data structures such as `dicts` or `lists`
  - create `functons` that are assigned to variables, used in expressions, and stored in different data structures.
  - understand and create` functions` that are/can be nested inside one another
- Understand that Python considers a function a form of  `callable object`. 
- Understand that a user-defined `function object` is created by a `function definition`.

</details>
<details>
<summary><b>Out of scope</b></summary>

- `named parameters` (_these can be touched on if needed_)
-  `default parameters` (_these can be touched on, if needed_)
- `arbitrary parameters`
- `*args & **kwargs`
- `keyword-only arguments`
- `/` and `*` for requiring parameter types
-  `functions-as-arguments` (_this can be mentioned, but shouldn't be required for the exercise_)
- `functions-as-returns`(_this can be mentioned, but will be covered in-depth in `higher-order functions`)
- `closures` (_these will be covered in a different exercise_)
- `decorators` (_these will be covered in a different exercise_)
-  `functools.wraps` (_this is used mostly for decorators_)
-  `functools` (_this will get its own exercise_)
- `comprehensions`
- `generators`
- `lambda`, `anonymous functions` (_these will be covered in a different exercise_)
- `recursion`
</details>

## Concepts

- `callable objects`
- `first-class functions`
- `global`
- `nested functions`
- `nonlocal`
- `return`, `implicit return`, `explicit return`
- `scope`
-  special `function attributes`
  
## Prerequisites

<details>
<summary><b>Proposed Prerequisites</b></summary>

_These are the concepts/concept exercises the student needs to complete/understand before solving this concept exercise.  Since `functions` is a a bit of a "meta" topic,  these can be adjusted to fit the parameters of the exercise, if needed._ 

- `basics`
- `bools`
- `comparisons`
- `lists`
- `list-methods`
- `loops`
- `numbers`
- `strings`
- `string-methods`
</details>

## Resources to Refer To

<details>
<summary><b>Links to resources</b></summary>

- [Python Docs: Naming and Binding](https://docs.python.org/3/reference/executionmodel.html#naming-and-binding)
- [Python Docs: Python Scopes and Namespaces](https://docs.python.org/3/tutorial/classes.html#python-scopes-and-namespaces)
- [Python Docs: Defining Functions](https://docs.python.org/3/tutorial/controlflow.html#defining-functions)
- [Function definitions (Python Library Reference)](https://docs.python.org/3/reference/compound_stmts.html#function-definitions)
- [Dan Bader:  Python s Functions are First-Class](https://dbader.org/blog/python-first-class-functions)
- [Real Python:  Python Scope & the LEGB Rule](https://realpython.com/python-scope-legb-rule/)
- [Stack Overflow: Short Description of Python's Scoping Rules?](https://stackoverflow.com/questions/291978/short-description-of-the-scoping-rules)
- [Real Python: Python Inner Functions: What are they Good For?](https://realpython.com/inner-functions-what-are-they-good-for/)
- [Stack Abuse: How to Use global and nonlocal variables in Python](https://stackabuse.com/how-to-use-global-and-nonlocal-variables-in-python/)
- [The `nonlocal` Statement (Python Docs)](https://docs.python.org/3/reference/simple_stmts.html#nonlocal)
- [The `global` Statement (Python Docs)](https://docs.python.org/3/reference/simple_stmts.html#global)
- [Real Python: The Python return Statement: Usage and Best Practices](https://stackabuse.com/how-to-use-global-and-nonlocal-variables-in-python/)
- [Calls (Python Library Reference)](https://docs.python.org/3/reference/expressions.html#calls)
- [Is it a Class or a Function?  It's a Callable! (Trey Hunner)](https://treyhunner.com/2019/04/is-it-a-class-or-a-function-its-a-callable/)
- [Callables in Python (Vimal A.R Blog)](https://arvimal.blog/2017/08/09/callables-in-python/)
- [Functions as Objects in Python](https://medium.com/python-pandemonium/function-as-objects-in-python-d5215e6d1b0d) -- _this is a subscription-based medium blog, so not good for hints or links_
- [Python Datamodel: Types](https://docs.python.org/3/reference/datamodel.html#types)
</details>
  
<br>

## Concept Files to Be Created

_Please see the following for more details on these files_:  [concepts](https://exercism.org/docs/building/tracks/concepts)

* ### `links.json`

  For more information, see [concept links file](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)

  -  The same resources listed in this issue can be used as a starting point for the [ `concepts/links.json`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)  file, if it doesn't already exist.
  -  If there are particularly good/interesting information sources for this concept that extend or supplement the concept exercise material & the resources already listed -- please add them to the `links.json` document.

* ### Concept `about.md`

  For more information, see [Concept `about.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-aboutmd)

  - This file provides information about this concept for a student who has completed the corresponding concept exercise.  It is intended as a reference for continued learning.  

* ### Concept `introduction.md`

  For more information, see [Concept `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-aboutmd)

  * This can also be a summary/paraphrase of the  `about.md` document listed above, and will provide a brief introduction of the concept for a student who has **not yet** completed the associated concept or practice exercises.   It should contain a good summation of the concept, but not go into lots of detail.

<br>

## Exercise Files to Be Created

* ### Exercise `introduction.md`

  For more information, see [**Exercise** `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docsintroductionmd)

  - This should also summarize/paraphrase  the above concept documents (either the `about` or the `introduction`), and can copy the concept introduction verbatim.   But the summary does need to have enough information and examples  for the student to complete all the tasks outlined for this concept exercise.
 
* ### Exercise `instructions.md`

   For more information, see [`instructions.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docsinstructionsmd)

   Instructions for an exercise usually center on a story that sets up the code challenge to be solved.  You can create your own story, or fork one from the ones listed [here](https://exercism.org/docs/building/tracks/stories).  Please make sure to give credit to the original authors if you use a story or fork an exercise.

* ### Exercise Exemplar Solution
  For more information, see [exemplar implementation](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-exemplar-implementation).
  
  This file should not use syntax or datas structures not introduced in this exercise or in this exercise's prerequisites.  It will be used as an "ideal" solution for the challenge, so make sure it conforms to PEP8 and other formatting conventions, and does not use single letter variable names.  It should also include proper module and function-level docstrings.  However, it should NOT include typehinting or type aliases.

* ### Exercise Stub for Implementation
  For more information, see [stub implementation](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-stub-implementation).
  
  This file should provide the expected function names imported for testing, and optionally TODO comments and or docstrings to aid the student in their implementation.  TODOs and docstrings are not required.

* ### Exercise Test Files
  For more information, see [Tests](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-tests).
  Additionally, please note that Python associates exercise tasks to tests via a [Pytest Marker](https://docs.pytest.org/en/7.1.x/example/markers.html), and uses [`unittest subtests`](https://docs.python.org/3/library/unittest.html#subtests) as a form of test paramaterization.  See the test file for [`Little Sisters Vocab`](https://github.com/exercism/python/blob/main/exercises/concept/little-sisters-vocab/strings_test.py) for examples of how these techniques work.

* ### Exercise Hints

  For more information on writing hints see [`hints.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docshintsmd)

  - Hints should provide enough information to get most students "un-stuck" and moving toward a solution.  They should **not** provide a student with a direct solution.
  - You can refer to one or more of the resources linked in this issue above, or analogous resources from a trusted source.  We prefer using links within the  [Python Docs](https://docs.python.org/3/) as the primary go-to, but other resources listed above are also good.  Please try to avoid paid or subscription-based links if possible.
 
* ### Exercise Metadata Files Under  `.meta/config.json`

  For more information on exercise  `.meta/`  files and formatting, see [concept exercise metadata files](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#metadata-files)

   * `.meta/config.json` - see [this link](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-metaconfigjson) for the fields and formatting of this file.
   * `.meta/design.md` - see [this link](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-metadesignmd) for the formatting of this file.  Please use the **Goal**, **Learning Objectives**,**Concepts**, **Prerequisites** and , **Out of Scope**  sections  from this issue.

<br>

## Exercise Metadata - Track

For more information on concept exercises and formatting for the Python track `config.json` , please see [`config.json`](https://exercism.org/docs/building/tracks/config-json).  The track `config.json` file can be found in the root of the Python repo.

You can use the below for the exercise **UUID**.  You can also generate a new  one via [exercism configlet](https://github.com/exercism/configlet), [uuidgenerator.net](https://www.uuidgenerator.net/version4), or any other favorite method.  The UUID must be a valid  [**V4 UUID**](https://en.wikipedia.org/wiki/Universally_unique_identifier).

- **Exercise UUID** : `9f38f70d-48f4-4934-ae6e-d8277ede8edc`
- **concepts** should be filled in from the Concepts section in this issue
- **prerequisites** should be filled in from the Prerequisites section in this issue  

  

## Implementation Notes

- As a reminder, code in the `.meta/examplar.py` file should **only use syntax & concepts introduced in this exercise or one of its prerequisite exercises.**  We run all our `examplar.py` files through PyLint, but do not strictly require module docstrings.  We **do** require function docstrings similar to [PEP257](https://www.python.org/dev/peps/pep-0257/).  See [this concept exercise `exemplar.py`](https://github.com/exercism/python/blob/main/exercises/concept/meltdown-mitigation/.meta/exemplar.py) for an example.
- Please **do not use** comprehensions, generator expressions, or other syntax not previously covered either in the introduction to this exercise, or to one of its prerequisites.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- In General, tests should be written using `unittest.TestCase` and the test file should be named `<EXERCISE-NAME>_test.py`.  

    - All asserts should contain a "user friendly" failure message (_these will display on the webiste to students, so be as clear as you can_). 
    - We use a `PyTest custom mark` to link test cases to exercise task numbers.  
    - We also use `unittest.subtest` to parameterize test input where/when needed.
Here is an [example testfile](https://github.com/exercism/python/blob/main/exercises/concept/making-the-grade/loops_test.py) that shows all three of these in action.

- While we do use [PyTest](https://docs.pytest.org/en/stable/) as our test runner and for some implementation tests, please check with a maintainer before using  a PyTest-specific test method, fixture,  or feature.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue, or contact one of the maintainers on our Slack channel.

**Comment by mukeshgurpude** [comment]
Hello @BethanyG, let me know if I can work on this.

**Comment by BethanyG** [comment]
@mukeshgurpude  YES!  And many thanks for stepping up and taking it on!  🎉     

**Bear with me:**  the links at the top for documentation should all be up-to-date, but the links down below in the detail are old, and need updating.  I will do that here shortly.  Please (_as always_) let me know if you have any questions or issues or need help.  

Also - gentle reminder to make sure you update and pull in all changes from upstream before getting a start.  I recommend rebasing `main` after doing a `fetch`, so that any introduced changes are inserted in order.  And -- depending on how you want to organize, you can do one PR for the `concept` (functions) and one for the concept `exercise`.  But that's up to you.  😄

**Comment by J08K** [comment]
Heya @mukeshgurpude, 
how's it going with this concept exercise? Haven't heard anything in a few days, so I'm just checking in!

Do you need me to look at some drafts of the exercise for you? Or do you have any questions? Feel free to ping me here, or submit a draft PR that we can look at and ping me there. You can also always ping me on slack! 

Good luck and happy coding! 🐍

**Comment by mukeshgurpude** [comment]
Hello @J08K,
actually, I'm in the middle of my examinations. I'll be working on this just after that.
I'll open a draft PR with my first edit only..

**Comment by J08K** [comment]
Cool, focus on your exams first ofcourse! Goodluck. Tag me in your draft PR when you make it. :)

**Comment by mukeshgurpude** [comment]
Hello, @BethanyG, @J08K
While #2623 is under review, I need some suggestions for the real world example for this exercise.
I thought of something as `interest-calculator`, which will just calculate the simple and compound interest, but in this it wasn't possible to show reusability🤔.

**Comment by bobahop** [comment]
@BethanyG  Any idea for what you want to see for the exercise here? Any analogous exercise you might want transcribed from another track? I can start looking at the tracks I'm familiar with. Nothing comes to mind right now.

**Comment by BethanyG** [comment]
@bobahop there are a few things to note here:

1.   This issue has very outdated links and descriptions and it also pings the attached users every time we make a comment, so I think its best if we close it in favor of a new issue using the new template.  I will do that shortly.   

2. I have a strong preference for original stories/exercises, unless the port/concept is an almost 1-to-1, as was the case with `bools`.  That **isn't** to say that porting is verboten, but I'd much rather have  a purpose-built exercise than a shoehorned one.  We have had trouble in the past with contributors trying to port something and then getting stalled because they were not able to transfer some core piece to Python.

3.   As I've mentioned, this exercise is quite far "down" the exercise tree. It should not be considered an exercise on how to _make_ functions in Python, but rather as a setup to both `higher order functions` and related concepts.  So we should try to think of something that deepens a student's understanding/use of functions, and exercises some of the features unique to Python.  Otherwise, this may as well be a practice exercise or a concept-only topic.

I'll add some thoughts/ideas to the new issue.

**Comment by BethanyG** [comment]
Closing this in favor of #3121
